### ISSUE TITLE
Vulkan: 2D Pixel Snap not working

### ISSUE THREAD

**Issue by markdibarry** [godotengine/godot#56793]
### Godot version

4.0dev (ce6de71)

### System information

Windows 10, NVIDIA GeForce GTX 960

### Issue description

I've seen other issues like this but they're all closed as fixed. I'm still having this issue. Pixels are not snapping correctly with either or both pixel snap options.
1 pixel gap in between collisions.

### Steps to reproduce

Create a project with a floor and characterbody2d with basic movement functions, gravity, etc.
Set display stretch mode to `viewport` and `keep`.
Enable `Snap 2d Transforms to Pixel` or `Snap 2d Vertices to Pixel` or both.

![image](https://user-images.githubusercontent.com/21325943/149568066-75a3685e-bda7-4da9-bd4f-dccd9f967dad.png)

### Minimal reproduction project

[PixelSnapBug.zip](https://github.com/godotengine/godot/files/7872520/PixelSnapBug.zip)

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot/issues/55006.

**Comment by AttackButton** [comment]
> Related to #55006.

Interesting enough the `smooth addon` https://github.com/lawnjelly/smoothing-addon was the only thing that completely fixed all this types of issues and jitterings in 3.x.

I'm having jittering issues again in 4.0 alpha 1, @lawnjelly any chance this being implemented by default on master too?

I'm working on a 2D platformer. Thanks for the work with this addon, @Calinou  and @lawnjelly.

**Comment by lawnjelly** [comment]
> I'm having jittering issues again in 4.0 alpha 1, @lawnjelly any chance this being implemented by default on master too?

We are currently adding fixed timestep interpolation (what the addon does) to core.
See #52846 for 3D in 3.x.
My plan is once 3D is merged and any bugs worked out, I'll then do 2D, then port to 4.x.

**Comment by svendixon** [comment]
It's been a while, hopefully this is on someones plate. 

If `Pixel 2D Transform to Pixel` or `Snap 2D Vertices to Pixel` or Both are set to `true` you will get those gaps. I'd also like to point out that if they are set to `false` sprites will get distorted and sprite atlases (with no padding) will also bleed as well when in motion (in the bellow example, this happens while moving in the y "jumping" since the sprite atlas has no padding at the bottom).

![pp_false](https://user-images.githubusercontent.com/14341036/227711887-1e638312-0bf7-4bfd-b37a-bc27d15835d7.png) ![pp_true](https://user-images.githubusercontent.com/14341036/227711895-5842aea5-1213-4e9d-8742-3e6f9ca3372c.png) ![pp_bleed](https://user-images.githubusercontent.com/14341036/227711899-bc32d08e-b596-4edc-b2c7-afd78dd69f58.png)

_A typical sprite atlas with no padding at the bottom_
![player](https://user-images.githubusercontent.com/14341036/227711305-28c855ed-688d-483f-a0e6-f7d3da6d149c.png)

So unlike Godot 3.x, this means at the moment Godot 4 has no solution for pixel perfect games. Hopefully it can be resolved soon. I am including a zip of this example project.
[PixelPerfect.zip](https://github.com/godotengine/godot/files/11068636/PixelPerfect.zip)

**Comment by svendixon** [comment]
I did a search for "is_snap_2d_transforms_to_pixel_enabled" and noticed that in `sprite_2d.cpp` it does a `floor()`, should this be a `round()` instead? For example if the value is 12.9999 you want it to be 13 not 12 no? 

I don't know how this will effect collisions for anything though since these are just sprite_2d and animated_sprite_2d. But this might fix an issue I had with odd number sized sprites? I had to give all odd number sized sprites a -0.5 offset.

_Line 102_
```gdscript
if (get_viewport() && get_viewport()->is_snap_2d_transforms_to_pixel_enabled()) {
    dest_offset = dest_offset.floor();
}
```
_Line 358_
```gdscript
if (get_viewport() && get_viewport()->is_snap_2d_transforms_to_pixel_enabled()) {
    ofs = ofs.floor();
}
```
And in `animated_sprite_2d.cpp`.

_Line 270_
```gdscript
if (get_viewport() && get_viewport()->is_snap_2d_transforms_to_pixel_enabled()) {
    ofs = ofs.floor();
}
```
I'm not a programmer and I don't know the inner workings of Godot so if I'm completely off base here, feel free to ignore this.

**Comment by huedev** [comment]
Are there any good workarounds to avoid this issue? With Godot 4.1 released and this issue still unresolved, it's challenging to work on pixel-perfect games right now, with seemingly no fix in sight.

**Comment by markdibarry** [comment]
@huedev You can do as you said, however, if your movement code doesn't work with integers, until 2D Pixel Snap is actually fixed, you should do this:
```gdscript
var float_position = 0

func _ready():
    float_position = global_position

func _physics_process(delta):
    global_position = float_position

    #your movement code here

    float_position = global_position
    global_position = global_position.round()
```

**Comment by huedev** [comment]
@markdibarry Finally had a chance to try this out, it seems to be working well! Since I'm using type hints I had to modify the initial value of `float_position` from 0 to `Vector2.ZERO`. Thanks for playing around with this and sharing your fix!

**Comment by markdibarry** [comment]
Fixed by #87297

**Comment by berarma** [comment]
I have this issue and fixed it applying the workaround in https://github.com/godotengine/godot/issues/56793#issuecomment-1872253285. Now I've tried Godot 4.3-dev5 that I've read should fix it but the workaround is still required. Nothing has changed.

I've tried to open the test project to see how it works but it gives many errors that I'm unable to fix.

I'm also overriding `_post_process_key_value` as a fast way to fix the position for sprites moved with an AnimationPlayer. This workaround is also still required with the latest snapshot.
